fud2/fud-core/src/run.rs

```diff
@@ -147,13 +162,66 @@ impl EmitBuild for OpEmitData {
             }
         }
 
-        emitter.build_cmd_with_vars(
-            outputs,
-            &self.op_name,
-            inputs,
-            &[],
-            &vars,
-        )?;
+        // Some outputs are created by user specified `shell_deps`. Others are not and have to be
+        // created by running every `shell`/`shell_deps` command.
+        let mut uncreated_outputs: HashSet<String> =
+            HashSet::from_iter((0..outputs.len()).map(|i| {
+                format!("${}", crate::run::io_file_var_name(i, false))
+            }));
+
+        // Write a sequence of rules for each cmd.
+        for (i, cmd) in self.cmds.iter().enumerate() {
+            let rule_name = format!("{}_rule_{}", self.op_name, i + 1);
+            emitter.rule(&rule_name, &cmd.cmd)?;
+            let targets = cmd.gens.iter().map(|s| s.as_str()).collect_vec();
+            let deps = cmd.deps.iter().map(|s| s.as_str()).collect_vec();
+            for &t in &targets {
+                uncreated_outputs.remove(t);
+            }
+            emitter.build_cmd_with_vars(
+                &targets,
+                &rule_name,
+                &deps,
+                &[],
+                &vars,
+            )?;
+        }
+
+        // Depend on all shell commands. The illusion is all shell commands are run in order, with
+        // the option to specify some shell commands as having explicit targets and dependancies as
+        // an optimization. They still all have to get run though.
+        let deps = inputs
+            .iter()
+            .cloned()
+            .chain(
+                self.cmds
+                    .iter()
+                    .map(|cmd| cmd.gens.last().unwrap().as_str()),
+            )
+            .collect_vec();
+
+        if uncreated_outputs.is_empty() {
+            // Even if all outputs have said to have been created by the user, all the calls to `shell`
+            // should still get run.
```

Comment:
I also don't really understand this… what does "said to have been created" mean?

Reply:
This goes back to my assertion that every rule created due to `shell` and `shell_dep` should always be run. It is possible for the output to be created by a `shell_dep` command but a `shell` command to follow that `shell_dep`.

The user says `shell_dep` constructs the output, but despite that, the following `shell` command must still be run. To do this, I generate a phony build rule generating a dummy output which notes all shell commands as dependencies.